central/reports/scheduler/schedule.go

```diff
@@ -411,6 +439,91 @@ func (s *scheduler) runPaginatedQuery(ctx context.Context, scopeQuery, cveQuery
+		// Current string query language doesn't support disjunction of multiple conjunctions, so we cannot build a
+		// string equivalent of deploymentsQuery for graphQL. Because of this, we first fetch deploymentIDs from
+		// deploymentDatastore using deploymentsQuery and then build string query for graphQL using those deploymentIDs.
+		scopeQuery := fmt.Sprintf("%s:%s", search.DeploymentID.String(),
+			strings.Join(deploymentIds[offset:mathutil.MinInt(offset+numDeploymentsLimit, len(deploymentIds))], ","))
+		r, err := s.execReportDataQuery(ctx, reportQueryPostgres, scopeQuery, cveQuery, paginatedQueryStartOffset)
+		if err != nil {
+			return r, err
+		}
+		resultData.Deployments = append(resultData.Deployments, r.Deployments...)
+		offset += numDeploymentsLimit
+	}
+	return resultData, nil
+}
+
+func (s *scheduler) execReportDataQuery(ctx context.Context, gqlQuery, scopeQuery, cveQuery string, offset int) (common.Result, error) {
+	response := s.Schema.Exec(ctx,
+		gqlQuery, "getVulnReportData", map[string]interface{}{
+			"scopequery": scopeQuery,
+			"cvequery":   cveQuery,
+			"pagination": map[string]interface{}{
+				"offset": offset,
+				"limit":  numDeploymentsLimit,
+				"sortOption": map[string]interface{}{
+					"field":    deploymentSortOption.GetField(),
+					"reversed": deploymentSortOption.GetReversed(),
+				},
+			},
+		})
+	if len(response.Errors) > 0 {
+		log.Errorf("error running graphql query: %s", response.Errors[0].Message)
+		return common.Result{}, response.Errors[0].Err
+	}
+	var res common.Result
+	if err := json.Unmarshal(response.Data, &res); err != nil {
+		return common.Result{}, err
+	}
+	return res, nil
+}
+
+func (s *scheduler) getDeploymentIDs(ctx context.Context, deploymentsQuery *v1.Query) ([]string, error) {
+	deploymentsQuery.Pagination = &v1.QueryPagination{
+		SortOptions: []*v1.QuerySortOption{
+			deploymentSortOption,
+		},
+	}
+	results, err := s.deploymentDatastore.Search(ctx, deploymentsQuery)
+	if err != nil {
+		return nil, err
+	}
+	return search.ResultsToIDs(results), nil
+}
+
+func groupByClusterAndNamespace(result common.Result) common.Result {
+	groupedDeployments := make([]*common.Deployment, 0, len(result.Deployments))
+	deploymentsByCluster := make(map[string][]*common.Deployment)
+	for _, deployment := range result.Deployments {
+		clusterName := deployment.Cluster.GetName()
+		deploymentsByCluster[clusterName] = append(deploymentsByCluster[clusterName], deployment)
+	}
+
+	for _, deployments := range deploymentsByCluster {
+		deploymentsByNamespace := make(map[string][]*common.Deployment)
+		for _, deployment := range deployments {
+			deploymentsByNamespace[deployment.Namespace] = append(deploymentsByNamespace[deployment.Namespace], deployment)
+		}
+		for _, deps := range deploymentsByNamespace {
+			groupedDeployments = append(groupedDeployments, deps...)
+		}
+	}
+
+	return common.Result{Deployments: groupedDeployments}
```

Comment:
Deployments IDs are fetch in the order of risk priority, but that order is lost while building the groups. Maps do not guarantee order.
Furthermore, I don't think you need maps. You can simply sort the `[]*common.Deployment` array.
```
	sort.SliceStable(deployments, func(i, j int) bool {
		if deployments[i].Cluster.GetName() == deployments[j].Cluster.GetName() {
			return deployments[i].Namespace < deployments[j].Namespace
		} 
		return deployments[i].Cluster.GetName() < deployments[j].Cluster.GetName()
	})
```

Reply:
Yeah, the global sort order will go away, but since we are iterating the sorted deployments and adding them to slices in  the map's value, the deployments will still be sorted within each group. This replicates the ordering that is in current vuln reports. I used maps for grouping to avoid sorting's nlogn time complexity but I agree that it affects the readability and the time for creating/deleting maps might probably be making it worse when there are large number of namespaces. So I am fine changing that to use sorting here.

That being said, I am in favor of removing the grouping altogether and just return deployments globally ordered by RiskPriority. We are sending the  results in csv format and the end user can group them if they want. I also don't see any documentation promising any ordering.

Reply:
Sorry, I think you misunderstood me. My comment was multi-fold:
- Is retrieving deployments by risk priority necessary? It is now ordering through this PR.
- The deployment ordering according to risk priority needs is not maintained during grouping, which made the effort to obtain deployments ordered by risk priority moot.
**- You can group the rows by sorting. The use of two maps seems unnecessary.**

Overall, I firmly believe we should not order csv rows by risk priority and keep the csv rows ordered by cluster+namespace combination because:
- Risk priority column is not included in the CSV.
- The report is a CSV report and not a deployment report. The data insights in csv are strictly focused on vulnerability and not the overall risk of the deployments.
- The readability of the csv drops dramatically without grouping. We do not care how customers manipulate the csv later, but the ACS output should be readable without sheet tools.

Reply:
Yes that makes sense and it simplifies the implementation, thanks for clarification. I would skip the ordering by riskPriority altogether and just sort by cluster+namespace since riskPriority is not a part of csv anyways.